File: 152-maximum-product-subarray/maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let n = nums.length
    let res = nums[0]
    let mx = nums[0]
    let mn = nums[0]
    for(let i=1;i<n;i++){
        if(nums[i]<0){
            let temp = mx
            mx=mn
            mn=temp
        }
        mx = Math.max(nums[i]*mx,nums[i])
        mn = Math.min(nums[i]*mn,nums[i])
        res = Math.max(res,mx)
    }
    return res
};

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        pre = 1
        prev_max = float('-inf')

        for i in nums:
            pre *=i
            prev_max=max(prev_max,pre)
            if pre == 0:
                pre =1
        print(prev_max)
        suff = 1
        suff_max = float('-inf')
        for i in range(len(nums)-1,-1,-1):
            suff*=nums[i]
            suff_max = max(suff_max,suff)
            if suff ==0:
                suff =1
        return max(suff_max,prev_max)